### PR TITLE
Daji/tweet parser

### PR DESCRIPTION
**Pull request by davidmjiang**
My first attempt at the tweet parser. It calls twitter api to get recent tweets that mention a particular account, then loops through them ,validates them and if they are valid, publishes them to our db. It's very hacky, and assumes location from geocoding and whole tweet body is the message. After getting the coordinates from the geocoding, it calls bing maps API to get the address. You can call /api/publishTweets?twitterId={accountId} (for SaveTheDose, account id is 1374256211961450501). I made a test tweet this morning mentioning SaveTheDose with location on and after calling the API, I can see it on the map: 
image

### CODE REVIEW

File: src/controller.js [hunk 1]
```diff
@@ -57,7 +58,7 @@ var API = {
     const options = {
       hostname: 'api.twitter.com',
       port: 443,
-      path: format('/2/users/{0}/mentions', twitterId),
+      path: format('/2/users/{0}/mentions?expansions=geo.place_id&place.fields=contained_within,country,country_code,full_name,geo,id,name,place_type', twitterId),
```
[the remark this answers is not in the document]
Comment (davidmjiang) [reply src/controller.js:61]: This is to ask for the place data, which doesn't come by default. https://developer.twitter.com/en/docs/twitter-api/data-dictionary/object-model/place

**Review by caiyuan1222 — APPROVED** [review]

**Comment by kgao** [comment]
Great to see the tweet parser! Can't wait to see more test examples for it!

**Review by kgao — COMMENTED** [review]
Mostly it looks fine, just need to review the little comment for the default values and verify some test cases for the parser (Exciting, maybe we can review it together in our next meeting!)

File: src/tweetparser.js [hunk 2]
```diff
@@ -0,0 +1,105 @@
+const API = require("./controller");
+
+const defaultRadius = 100;
+const defaultDuration = 7200000 // 2 hours in ms
```
[the remark this answers is not in the document]
Comment (davidmjiang) [reply src/tweetparser.js]: moved

File: src/controller.js [hunk 3]
```diff
@@ -134,6 +134,32 @@ var API = {
...
+      path: format('/2/tweets/search/stream'),
+      method: 'GET',
+      headers: {
+        'Authorization': format('Bearer {0}', twitterBearToken)
+      }
+    }
+    const req = https.request(options, (res) => {
+      console.log(`statusCode: ${res.statusCode}`)
+      let data = ''
+      res.on('data', (chunk) => {
```
Comment (davidmjiang) [inline src/controller.js]: looks good. I think there are two things missing for this to work e2e. 
1. Here, we need to read the chunk and if it's a valid tweet, pass it on to the tweet parser for processing. See https://github.com/twitterdev/real-time-tweet-streamer/blob/master/server/server.js#L122 for an example. 
2. We need to start the stream once when the server starts up. We don't want to have to manually call it from postman.

Comment (davidmjiang) [reply src/controller.js:150]: @kgao @caiyuan1222 @wangmeng255 I made these changes with latest push. It starts the stream when the server starts and passes any new tweets to the parser, which will publish them if they are valid. To test, pull the latest, run the server (npm run start) then make a tweet mentioning @SaveTheDose with location on. Then refresh the dashboard (aka.ms/commoncircle) and the new dot should show up in the location of the tweet

File: routes.js [hunk 4]
```diff
@@ -143,12 +143,6 @@ router.get('/api/getRules', function(req, res) {
   });
 });
 
-router.get('/api/getTweets', function(req, res) {
```
Comment (davidmjiang) [inline routes.js]: don't need to call this anymore, we start it when the server starts

File: src/controller.js [hunk 5]
```diff
@@ -134,11 +134,11 @@ var API = {
     })
     req.end()
   },
-  getNewTweets: function (cb) {
+  getNewTweets: function (tweetParser) {
     const options = {
       hostname: 'api.twitter.com',
       port: 443,
-      path: format('/2/tweets/search/stream'),
+      path: format('/2/tweets/search/stream?expansions=geo.place_id&place.fields=contained_within,country,country_code,full_name,geo,id,name,place_type'),
```
Comment (davidmjiang) [inline src/controller.js:141]: adding this to get place data included

**Pull request merged**